### PR TITLE
Usr/alexanders/build real config

### PR DESCRIPTION
**Pull request by sanyaz-s**
Please, do the review. 
Here was done:
1. Add mechanism, which inserts real config params into original params
2. Add log mechanism into the project

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +0,0 @@
-# pap_project
```
Comment (AlexLostOwl) [inline README.md]: Please return readme file

File: behave.ini [hunk 2]
```diff
@@ -1,4 +1,10 @@
...
-device         = ws_2003
-appliance_ram  = 2
-appliance_size = 8
\ No newline at end of file
+
+#machine.virtual.-0.t52345426 = 123
+machine.virtual.-0.ip = 8.8.8.8
+machine.virtual.-0.username = pikachu
+;appliance.virtual.ssh.username = charmonder
+;Device = 'windows
```
Comment (AlexLostOwl) [inline behave.ini]: Let's not put here examples. Maybe will be better put examples into the readme?

File: environments/machine.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+virtual:
+-
+    ip: &device_host1
```
Comment (AlexLostOwl) [inline environments/machine.yaml]: Let's not call machines as device

File: features/environment.py [hunk 4]
```diff
@@ -1,5 +1,14 @@
...
-    userdata = context.config.userdata
-    print(f'Device is: {userdata["device"]} system')
-    print(f'Applinace RAM is: {userdata["appliance_ram"]} GB')
-    print(f'Applinace size is: {userdata["appliance_size"]} GB')
+    print(ConfigBuilder(context.config.userdata).machine())
+    # print(ConfigBuilder(context.config.userdata).appliance())
+    # print(ConfigBuilder(context.config.userdata).esxi())
+    # print(ConfigBuilder(context.config.userdata).general())
+    # print(ConfigBuilder(context.config.userdata).jenkins())
+    log.info('Configs are already in before_all!')
```
Comment (AlexLostOwl) [inline features/environment.py]: Do we really need here print?

File: lib/helpers/yaml_loader.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        with open(path_to_file, 'r') as stream:
+            try:
+                file_name = self._get_name_of_dictionary(path_to_file)
+                dictionary = {file_name: yaml.load(stream, Loader=yaml.FullLoader)}
+            except yaml.YAMLError as exc:
+                self.union_raise_log(f'{exc}')
+            finally:
+                log.info(f'Yaml file - {path_to_file} is successfully read!')
+
+            return dictionary
```
Comment (AlexLostOwl) [inline lib/helpers/yaml_loader.py]: Space?

File: lib/helpers/config_builder.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        log.info(f'Insert value "{str(value)}" into the default yaml config is successful')
+
+        return configs_data                           # unnecessary return, but let's return result of configs_data
+
+    @staticmethod
+    def check_last_key_in_config(last_key, configs_data):
+        if not (isinstance(configs_data, dict)):
+            YamlLoader.union_raise_log(f'It has not been a proper config already :( : {str(configs_data)}. '
+                                       f'Maybe, do you confused in variable types? This is not a dict!')
+        elif last_key not in configs_data.keys():
```
Comment (AlexLostOwl) [inline lib/helpers/config_builder.py]: elif without else?

File: lib/helpers/config_builder.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                if isinstance(configs_data, dict):
+                    YamlLoader.union_raise_log(f'The config: {str(configs_data)} + has not any such keys: "{str(key)}"')
+                elif isinstance(configs_data, list):
+                    YamlLoader.union_raise_log(f'The index : "{str(key)}" of array {str(configs_data)} is outside')
+                else:
+                    YamlLoader.union_raise_log(f'The type of config must be in [dict, list] vs {exc}')
+        ConfigBuilder.check_last_key_in_config(keys[-1], configs_data)
+        configs_data[keys[-1]] = value                # take dict from last key and equal it to value
+        log.info(f'Insert value "{str(value)}" into the default yaml config is successful')
+
```
Comment (AlexLostOwl) [inline lib/helpers/config_builder.py]: space?

File: lib/helpers/config_builder.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        return configs_data
+
+    # mix string from USER DATA with configs
+    # insert our value into result dictionary
+    @staticmethod
+    def _insert_value_into_config(configs_data, keys, value):
+        for key in keys[:-1]:                         # take key up to last, equal to [0:last - 1]
+            try:
+                configs_data = configs_data[key]      # nest dict with key into dict
+            except Exception as exc:
```
Comment (AlexLostOwl) [inline lib/helpers/config_builder.py]: Do we really need to catch all exceptions?

File: lib/helpers/yaml_loader.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        file_names = []
+        all_dicts = {}
+        try:
+            file_names = [file for file in os.listdir(path_to_dir) if
+                          file.endswith(".yaml")]                       # cross platform getting file_names from dir
+        except Exception as exc:
+            self.union_raise_log(f'{exc}')
+        finally:
+            log.info('All yaml files are ready to read!')
+
```
Comment (AlexLostOwl) [inline lib/helpers/yaml_loader.py]: I'm not sure about spaces here. Is there any best practice guide for this?

File: requirements.txt [hunk 10]
```diff
@@ -1 +1,6 @@
+# sudo apt-get install libpcap-dev libpq-dev or install update of setup.py
+# pip install pyyaml
+# https://yaml.readthedocs.io/en/latest/install.html - DO IT BEFORE insatall YAML
```
Comment (AlexLostOwl) [inline requirements.txt]: Do we really need this?will pip install -r requirements.txt will cover?

File: results/log.log [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+2019-11-04 20:43:39,697 - yaml_loader.py - INFO - Yaml file - C:\Users\MALE\PycharmProjects\pap_project\environments\network_settings.yaml is successfully read!
+2019-11-04 20:43:39,702 - yaml_loader.py - INFO - Yaml file - C:\Users\MALE\PycharmProjects\pap_project\environments\psa.yaml is successfully read!
+2019-11-04 20:43:39,703 - yaml_loader.py - INFO - Yaml file - C:\Users\MALE\PycharmProjects\pap_project\environments\template_storage.yaml is successfully read!
+2019-11-04 20:43:39,705 - yaml_loader.py - INFO - Yaml file - C:\Users\MALE\PycharmProjects\pap_project\environments\vcenter.yaml is successfully read!
+2019-11-04 20:43:39,705 - config_builder.py - INFO - Insert value "8.8.8.8" into the default yaml config is successful
+2019-11-04 20:43:39,705 - config_builder.py - INFO - Insert value "pikachu" into the default yaml config is successful
+2019-11-04 20:43:39,705 - config_builder.py - INFO - Insert value "charmonder" into the default yaml config is successful
+2019-11-04 20:43:39,705 - config_builder.py - INFO - Merge user data with default yaml configs is successful
+2019-11-04 20:43:39,705 - config_builder.py - INFO - Configs are already in before_all!
```
Comment (AlexLostOwl) [inline results/log.log]: please add log into gitignore

File: tools/logger.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    def __init__(self, exec_file):
+        super().__init__()
+        self.__exec_file = exec_file
+        self.__path_to_logs = 'results/log.log'
+
+    def new(self):
+        self.__create_file_log()
+        logger = logging.getLogger(self.__exec_file)
+        handler = logging.FileHandler(self.__path_to_logs)
+        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
```
Comment (AlexLostOwl) [inline tools/logger.py]: f string isn't work here?

File: README.md (hunk 1 above)
Comment (sanyaz-s) [reply README.md]: +, reterned

File: environments/machine.yaml (hunk 3 above)
Comment (sanyaz-s) [reply environments/machine.yaml]: +

File: behave.ini (hunk 2 above)
Comment (sanyaz-s) [reply behave.ini]: yes, it's really cooler

File: lib/helpers/yaml_loader.py (hunk 5 above)
Comment (sanyaz-s) [reply lib/helpers/yaml_loader.py]: Spaces are separated big pieces of code, I think they need and it's good practice

File: lib/helpers/config_builder.py (hunk 8 above)
Comment (sanyaz-s) [reply lib/helpers/config_builder.py]: Yes, we need them for catching exceptions when we have some errors in config params.

File: lib/helpers/yaml_loader.py (hunk 9 above)
Comment (sanyaz-s) [reply lib/helpers/yaml_loader.py]: Spaces are separated big pieces of code, I think they need and it's good practice

File: features/environment.py (hunk 4 above)
Comment (sanyaz-s) [reply features/environment.py]: No, sorry, I have deleted it.

File: requirements.txt (hunk 10 above)
Comment (sanyaz-s) [reply requirements.txt]: No, It is deleted now

File: results/log.log (hunk 11 above)
Comment (sanyaz-s) [reply results/log.log]: Log is in git ignore in Folder Results

File: lib/helpers/config_builder.py (hunk 8 above)
Comment (sanyaz-s) [reply lib/helpers/config_builder.py]: For example: The config: {str(configs_data)} + has not any such keys: "{str(key)}"

File: lib/helpers/config_builder.py (hunk 6 above)
Comment (sanyaz-s) [reply lib/helpers/config_builder.py]: Yes, I have done it.

File: tools/logger.py (hunk 12 above)
Comment (sanyaz-s) [reply tools/logger.py]: f string doesn't work here.

File: behave.ini [hunk 13]
```diff
@@ -1,4 +1,10 @@
+# example: "machine.Device". Symbol "." - for access to level dictionary /pap_project/environments/<name before symbol.>
+# symbol "-" for access to level of array
+
 [behave.userdata]
```
[the remark this answers is not in the document]
Comment (sanyaz-s) [reply behave.ini]: Да, так и есть, в переделанном варианте так и есть

File: tools/logger.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import logging
+import os
+
+
+class Logger:
```
[the remark this answers is not in the document]
Comment (sanyaz-s) [reply tools/logger.py]: Обсудим

File: behave.ini [hunk 15]
```diff
@@ -1,4 +1,10 @@
...
-device         = ws_2003
-appliance_ram  = 2
-appliance_size = 8
\ No newline at end of file
+
+#machine.virtual.-0.t52345426 = 123
+machine.virtual.-0.ip = 8.8.8.8
```
[the remark this answers is not in the document]
Comment (sanyaz-s) [reply behave.ini]: Такая структура

File: lib/helpers/config_builder.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
+import os
+from lib.helpers.yaml_loader import *
+from tools.logger import Logger
+
+log = Logger(os.path.basename(__file__)).new()
+
+
+class ConfigBuilder(YamlLoader):
```
[the remark this answers is not in the document]
Comment (sanyaz-s) [reply lib/helpers/config_builder.py]: Обсудим))

**Pull request merged**